logstash-core/lib/logstash/agent.rb

```diff
@@ -464,51 +436,27 @@ def update_metrics(converge_result)
 
   def update_success_metrics(action, action_result)
     case action
-      when LogStash::PipelineAction::Create
-        # When a pipeline is successfully created we create the metric
-        # place holder related to the lifecycle of the pipeline
-        initialize_pipeline_metrics(action)
       when LogStash::PipelineAction::Reload
         update_successful_reload_metrics(action, action_result)
     end
   end
 
   def update_failures_metrics(action, action_result)
-    if action.is_a?(LogStash::PipelineAction::Create)
-      # force to create the metric fields
-      initialize_pipeline_metrics(action)
-    end
-
-    @instance_reload_metric.increment(:failures)
-
-    @pipeline_reload_metric.namespace([action.pipeline_id, :reloads]).tap do |n|
-      n.increment(:failures)
-      n.gauge(:last_error, { :message => action_result.message, :backtrace => action_result.backtrace})
-      n.gauge(:last_failure_timestamp, LogStash::Timestamp.now)
-    end
-  end
 
-  def initialize_agent_metrics
-    @instance_reload_metric.increment(:successes, 0)
-    @instance_reload_metric.increment(:failures, 0)
-  end
+    @witness_reloads.failure
 
-  def initialize_pipeline_metrics(action)
-    @pipeline_reload_metric.namespace([action.pipeline_id, :reloads]).tap do |n|
-      n.increment(:successes, 0)
-      n.increment(:failures, 0)
-      n.gauge(:last_error, nil)
-      n.gauge(:last_success_timestamp, nil)
-      n.gauge(:last_failure_timestamp, nil)
-    end
+    witness_pipeline_reloads = @witness_pipelines.pipeline(action.pipeline_id).reloads
+    witness_pipeline_reloads.failure
+    witness_pipeline_reloads.error.message(action_result.message)
+    witness_pipeline_reloads.error.backtrace(action_result.backtrace.to_s)
+    witness_pipeline_reloads.last_failure_timestamp(LogStash::Timestamp.now)
```

Comment:
Can we consolidate these operations into a single method on the witness? My thinking is that it'd be a cleaner separation of concerns to have the externally used witness methods correspond to lifecycle events rather than act as setters.

Maybe something like:

```ruby
witness.pipeline(action.pipeline_id).failure(message, action_result.backtrace.to_s)
```

Reply:
+1 , i will gather these all suggestions like this up over at https://github.com/elastic/logstash/issues/7973 and make a final sweep of the code with those changes.